accord-core/src/main/java/accord/local/Command.java

```diff
@@ -18,1178 +18,1435 @@
-            throw new IllegalStateException();
-
-        if (executeAt() == null)
+        @Override
+        public String toString()
         {
-            TxnId txnId = txnId();
-            // unlike in the Accord paper, we partition shards within a node, so that to ensure a total order we must either:
-            //  - use a global logical clock to issue new timestamps; or
-            //  - assign each shard _and_ process a unique id, and use both as components of the timestamp
-            // if we are performing recovery (i.e. non-zero ballot), do not permit a fast path decision as we want to
-            // invalidate any transactions that were not completed by their initial coordinator
-            if (ballot.equals(Ballot.ZERO)) setExecuteAt(safeStore.preaccept(txnId, partialTxn.keys()));
-            else setExecuteAt(safeStore.time().uniqueNow(txnId));
+            return "ListenerProxy{" + listenerId + '}';
+        }
 
-            if (status() == NotWitnessed)
-                setStatus(PreAccepted);
-            safeStore.progressLog().preaccepted(this, shard);
+        public TxnId txnId()
+        {
+            return listenerId;
         }
-        else
+
+        @Override
+        public void onChange(SafeCommandStore safeStore, TxnId txnId)
         {
-            // TODO (expected, ?): in the case that we are pre-committed but had not been preaccepted/accepted, should we inform progressLog?
-            setSaveStatus(SaveStatus.enrich(saveStatus(), DefinitionOnly));
+            Commands.listenerUpdate(safeStore, safeStore.command(listenerId), safeStore.command(txnId));
         }
-        set(safeStore, Ranges.EMPTY, coordinateRanges, shard, route, partialTxn, Set, null, Ignore);
 
-        notifyListeners(safeStore);
-        return AcceptOutcome.Success;
+        @Override
+        public PreLoadContext listenerPreLoadContext(TxnId caller)
+        {
+            return PreLoadContext.contextFor(Utils.listOf(listenerId, caller), Keys.EMPTY);
+        }
     }
 
-    public boolean preacceptInvalidate(Ballot ballot)
+    public static CommandListener listener(TxnId txnId)
     {
-        if (promised().compareTo(ballot) > 0)
-        {
-            logger.trace("{}: skipping preacceptInvalidate - witnessed higher ballot ({})", txnId(), promised());
-            return false;
-        }
-        setPromised(ballot);
-        return true;
+        return new Listener(txnId);
     }
 
-    public AcceptOutcome accept(SafeCommandStore safeStore, Ballot ballot, PartialRoute<?> route, Seekables<?, ?> keys, @Nullable RoutingKey progressKey, Timestamp executeAt, PartialDeps partialDeps)
+    private abstract static class AbstractCommand extends Command
     {
-        if (this.promised().compareTo(ballot) > 0)
+        private final TxnId txnId;
+        private final SaveStatus status;
+        private final Status.Durability durability;
+        private final RoutingKey homeKey;
```

Comment:
To clarify, other commands wouldn't need it because they should have a route which already has a home key?

Reply:
Right, and we don’t need to maintain it at all if we’re more than NotWitnessed. But I think it would in general potentially be good to have NotWitnessed be a separate thing

Reply:
What about an Unwitnessed route implementation that only has a home key for that case? It looks like we could move the progressKey in there as well, since it’s only updated alongside the homeKey. That would remove unneeded attributes from the other classes without complicating CommonAttributes.Mutable and/or the places that use it.

Since we no longer need an EMPTY command, we could collapse most of AbstractCommand into Command, with status and durability moved into PreAccepted, with NotWitnessed extending Command.

Reply:
Oh, hmm. I'm not sure about complicating the idea of a Route, as it is used quite widely in the codebase. Having to worry about not-really-Routes could be complicating elsewhere.

Reply:
Should homeKey be removed from Route then? It seems odd that the route contains a homeKey some of the time, and other times it’s just an attribute on the command

Reply:
I don’t really follow? It’s possible to know the homeKey and not the Route, but it shouldn’t be possible to know the Route and not the homeKey

A Route is incomplete without a homeKey, as the whole concept of a Route is a homeKey+Unseekables, and the point of the Route is to declare where we expect both to be known. Again, they’re used throughout the codebase, so would be quite a disruptive change and potentially lead to mistakes with one being known and not the other.

Reply:
is it possible for invalidated commands to have a homeKey, but no route?

Reply:
So in that case, I’d say we should leave things as they are. AcceptedInvalidate and Invalidate use Accepted and Committed classes respectively, so they also need to support homeKey only and Route. It may make sense to introduce an additional Invalidated command class, though I’m not sure if it would cause problems in scenarios where an invalidation was accepted, then the command becomes committed after having forgotten it’s definition. Either way though, I think exploring that should be follow on work.

Reply:
Fair. I think probably we should consider removing the independent update of homeKey altogether, as I don’t think it was necessary. But agree it should be follow up work to clean this up further, it’s a minor thing.